src/abstract_charm.py

```diff
@@ -248,8 +250,10 @@ def reconcile(self, event=None) -> None:  # noqa: C901
                 ):
                     self._database_provides.reconcile_users(
                         event=event,
-                        router_read_write_endpoint=self._read_write_endpoint,
-                        router_read_only_endpoint=self._read_only_endpoint,
+                        router_read_write_endpoint=self._read_write_endpoint(),
+                        router_read_only_endpoint=self._read_only_endpoint(),
+                        exposed_read_write_endpoint=self._read_write_endpoint(is_internal=False),
+                        exposed_read_only_endpoint=self._read_only_endpoint(is_internal=False),
```

Comment:
if this change is made in abstract_charm, then this will need to be propagated to the vm charm, and i am not sure that the `exposed_read_write_endpoint` and `exposed_read_only_endpoints` make sense in the vm context. can the mux-ing of these 2 values (router_read_write_endpoint vs. exposed_read_write_endpoint, and router_read_only_endpoint vs exposed_read_only_endpoint) be done in _read_write_endpoint() and _read_only_endpoint()?

i think that as i write this, i realize that the k8s charm is different than the vm charm in that k8s can be related to multiple clients while vm is subordinated and so only related to one client

Reply:
@carlcsaposs-canonical @shayancanonical (@paulomach @taurus-forever fyi) to clarify: right now the only option is locally calling mysql-router. That happens [because we are setting `mysql_client` as a `container` type relation](https://github.com/canonical/mysql-router-operator/blob/5d89fb6acd08ec3ea2e65356fad520ad5d9862d9/metadata.yaml#L24). If we remove that `container`, I don't think Juju will automatically deploy the mysql-router as a subordinate of the app charm.

We cannot either deploy the VM charm in the same as the mysql itself, [as snap does support multiple instances of the same snap, but this is still deemed experimental](https://snapcraft.io/docs/parallel-installs).

Now, there is a need here for an stand-alone mysql-router in VM as well, just like k8s charm, so we can connect non-charmed clients to the router as well. In this case, it will eventually make sense to separate internal (i.e. running as subordinate of a given client) vs. external (i.e. running somewhere else outside of the same client VM). 

For the VM case, my proposal is to add a 2nd mysql_client relation, which is not `container`. This way, a given mysql-router can be a subordinate for one type of app and still receive remote relations from other apps. This way, we could deploy with mysql-router connected with a dummy "ubuntu" charm, so we build them as gateways effectively to the mysql DB. If an user does not specify if it wants the `database-subordinate` (example of new name for the container-type relation) or the `database-standalone`, then Juju will complain that we have more than one mysql_client interfaces that are options and fail, forcing the user to select one of them in the end.

Wdyt?

Reply:
Another alternative: if the VM mysql-router is the subordinate of a given app that passes the `external-node-connectivity`, then that VM mysql-router *also opens* ports for remote connections. I think this is far simpler!

Then, we could have the data-integrator as the "gateways" in this case, mysql-router being related and subordinated to these integrator units.

Reply:
actually, the second comment is precisely head-on! i am currently working on https://warthogs.atlassian.net/browse/DPE-3689 which is exactly an integration with data-integrator which passes the `external-node-connectivity` flag. i originally wrote the comment when i discovered the divergence in the file's logic when working on the vm charm